### ISSUE TITLE
Add bidi / RTL support

### ISSUE THREAD

**Issue by ahangarha** [LemmyNet/lemmy-ui#347]
### Is your proposal related to a problem?

Considering the very nature of social media platforms (like Lemmy), people with different languages would create content. It becomes problematic when one creates content in RTL languages (like Persian, Arabic, Urdu,...). Lemmy is not ready to render RTL (or even mixed RTL/LTR) text properly.

### Describe the solution you'd like

There are two approaches in general. One is through  HTML modification which can be done by add `dir="auto"` attribute to elements containing user generated text or by wrapping it in `<bdi>` tags. The other approach would be through CSS by adding `unicode-bidi: plaintext`. In either of approaches, there should be avoidance in using absolute positioning or spacing. For example instead of `text-align: left` we should use `text-align: start` or instead of `padding-left` we should use `padding-inline-start` unless there be some strong UX reason.

### Describe alternatives you've considered

--

### Additional context

Just keep in mind that bidi (bidirectional text) is different from using RTL or LTR in general based on the language user choose for the interface. 

I have opened this issue here not on back-end because I think this can be managed on front-end.

**Comment by dessalines** [comment]
I have no familiarity with this, but anyone should feel free to give it a try.

**Comment by Nutomic** [comment]
I found this page which gives a lot ofdetails on how bidirectional text works: https://www.w3.org/International/articles/inline-bidi-markup/

**Comment by cadars** [comment]
Another great resource: https://rtlstyling.com/posts/rtl-styling

**Comment by ahangarha** [comment]
> Another great resource: https://rtlstyling.com/posts/rtl-styling

Please notice that bidi is not about RTL. One can use RTL if it is known that the content is RTL. We are talking about a social media in which people may write text in different languages and even wrote mixed content. We don't have and we don't want to have rich text editor in which the user can specify which direction should be applied on a specific portion of the text. We want to handle this situation automatically which is called bidirectional text support.

In most of the cases, adding `dir="auto"` to the element which contains the text is the trick. In addition we need to use other values that absolute left|right for text alignment or padding/margin or even borders.

**Comment by cadars** [comment]
> In most of the cases, adding dir="auto" to the element which contains the text is the trick.

Yes, to **each** element if you want true mixed content support. (And to the top element only for nested elements like `ul` or `blockquote` https://codepen.io/cadars/pen/poeXMXZ )

I did a quick review of what would be needed to fully support RTL languages in lemmy-ui:

1. Editor: add dir="auto" to all elements users can create, like @ahangarha suggested.
2. Add the proper `lang` attribute to the `html` element when switching the language. For now it's always `lang="en"`.
3. Add `dir="rtl"` to the `html` element for RTL languages.
4. Switch to logical CSS properties:
 - Because lemmy-ui is using flexbox for layout at its core this is quite easy, see https://github.com/cadars/lemmy-ui/commit/4ded16218bd977787d9b7ea0f8d19b09a294f299 https://github.com/cadars/lemmy-ui/commit/025a058120df457f053af6a51b7a7c01902753e1
 - The hard part: most of the styling comes from the (bootstrap) themes CSS, and rewriting each theme by hand, or asking future theme writers to only use logical properties isn't really an option: a solution would be to use [RTLCSS](https://rtlcss.com/) or similar(?) to convert each theme to their RTL counterpart. (of course this adds yet another dependency to the project…)

**Comment by ahangarha** [comment]
Things like RTLCSS wouldn't help as per my understanding. We need to modify certain styles. For example:
- `text-align:left|right` -> `text-align:start|end`
- `padding-left|right:*` -> `padding-inline-start|end`
- `margin-left|right:*` -> `margin-inline-start|end`
- and somehow similar properties for border,...

**Comment by cadars** [comment]
In an ideal world, yes, but lemmy-ui uses Bootstrap themes, and rewriting them is no small feat; so using a “converter”, while being a less elegant solution, sounds way more practical.

**Comment by ahangarha** [comment]
Let's say some bidifix. I don't know when I would find time to work on this, but it should easy to implement.

**Comment by dessalines** [comment]
Whoever takes this on, I want to mention that lemmy-ui uses bootstrap, and it has RTL instructions here: https://getbootstrap.com/docs/5.0/getting-started/rtl/

**Comment by ahangarha** [comment]
> Whoever takes this on, I want to mention that lemmy-ui uses bootstrap, and it has RTL instructions here: https://getbootstrap.com/docs/5.0/getting-started/rtl/

Again, this is for explicitly define direction. Such solutions won't work for us. We need dynamic directions based on the content.

**Comment by jayabie** [comment]
Any update on this case guys?
We need the RTL ASAP please.

**Comment by dessalines** [comment]
PRs adding this welcome.

**Comment by ahangarha** [comment]
I checked dependencies. Lemmy uses `markdown-it` for handling markdown. I have success in convincing `marked` to add bidi support (they made a plugin for it) but still struggling with developers on `markdown-it`.

There are some workaround. If lemmy developers are ok with adding this feature in two steps, I think we can start working on it. We can make markdown sections to follow the first character's direction till `markdown-it` add bidi support or a plugin be made for it. then we can remove the relevant modification for these section since it will be fully handled by markdown parser.

If it is ok, I try to find some time to create local development environment.

**Comment by dessalines** [comment]
Sweet, let me know if you need any help setting up a development environment.

**Comment by ahangarha** [comment]
In few week I will come back again. First I try to do it on my own and if I failed, I will seek help.

**Comment by ahangarha** [comment]
I just wanted to inform you that we are working on a plugin for `markdown-it` which is used in this project for parsing md to html. This plugin called [`markdown-it-bidi`](https://github.com/ahangarha/markdown-it-bidi/) and is under development. not ready to be used in production but I think it is on the track.

I decided to do this so that not only we can help lemmy to fix this issue but many more projects can benefit from. It is under LGPL which I think is very good :). So if I didn't follow up here, that was the reason. Hopefully in sometimes we can add that in this project and...

If anyone interested, please check the code and share your thoughts.

**Comment by jayabie** [comment]
@dessalines @ahangarha Friends, no progress with this yet?

**Comment by ahangarha** [comment]
@jayabie I am looking for an opportunity to open some time for it.

**Comment by jayabie** [comment]
@dessalines @ahangarha 

No update my friends?

**Comment by ahangarha** [comment]
@jayabie I am going to try now.
As long as there is no nested elements (like list within list) the plugin is usable even now.

But let me try it on the actual project in development environment. If successful, I send PR. Would it be fine? I mean within 2 days I might ask to assign this issue to me. Would it be fine?

**Comment by jayabie** [comment]
Hello @ahangarha 

I'm just a guy seeking for RTL, i think you should ask @dessalines if any extra permissions needed.

**Comment by Nutomic** [comment]
> But let me try it on the actual project in development environment. If successful, I send PR. Would it be fine? I mean within 2 days I might ask to assign this issue to me. Would it be fine?

There is no rush, no one else is working on this. Just work on it as you can, and make a pull request when you are ready. Also feel free to ask if anything is unclear.

**Comment by ahangarha** [comment]
@Nutomic I have issue in building backend. How can I get help? I don't think here is a good place.

**Comment by Nutomic** [comment]
@ahangarha What error do you get? Make sure to run `git submodule init && git submodule update` before build. And follow the documentation.

https://join-lemmy.org/docs/en/contributing/contributing.html

**Comment by ahangarha** [comment]
This was and still is the error:
```
➜  lemmy git:(main) cargo check
   Compiling lemmy_utils v0.16.5 (/****/lemmy/crates/utils)
    Checking futures v0.3.21
    Checking actix-server v2.1.1
    Checking opentelemetry v0.17.0
error: failed to run custom build command for `lemmy_utils v0.16.5 (/****/lemmy/crates/utils)`

Caused by:
  process didn't exit successfully: `/****/lemmy/target/debug/build/lemmy_utils-53c738f32a60fcf3/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=translations/email/en.json
  cargo:rerun-if-changed=translations/email/pt.json
  cargo:rerun-if-changed=translations/email/ko.json
  cargo:rerun-if-changed=translations/email/fi.json

  --- stderr
  Error: Fmt(Os { code: 2, kind: NotFound, message: "No such file or directory" })
warning: build failed, waiting for other jobs to finish...
error: build failed
```
It seems it cannot get translations.

**Comment by Nutomic** [comment]
So you are missing the submodule. Did you run the commands from my previous comment? And what does git status show?

**Comment by ahangarha** [comment]
Yes. I ran.

I ran clone with --recursive and also I ran the commands you gave me. Nothing happened. see:

```
➜  lemmy git:(main) git submodule init && git submodule update
➜  lemmy git:(main)
```

**Comment by dessalines** [comment]
We also saw issues with that when people were running old versions of rust, or not properly using `rustup` to get new rust versions.

If the files are there, make sure you're at least on rust `1.60`.

What does `rustc --version` say?

**Comment by dessalines** [comment]
Also lmk if I can help in any way... we desperately need RTL, but I have no experience with it, how it should work, or how to test it.

**Comment by ahangarha** [comment]
My rust version is `1.59.0`.

For RTL (I would say bidi not RTL) it depends. I had a presentation about this in LinuxAppSummit 2022. I have to prepare its video and publish. Otherwise, we need to spend some time to work on it. Sometimes it gets tricky.

**Comment by dessalines** [comment]
Why is that submodule on such an old commit.

run `git submodule update --remote`

If that doesn't work, delete the folder and run this exactly:

`git clone https://github.com/lemmynet/lemmy --recurse-submodules --remote-submodules`

**Comment by ahangarha** [comment]
I ran the first command. It made some changes but didn't fix the issue.

I removed the repo and made new clone with the second command but again the same issue. These are the last few lines of the result:

```
    Checking bcrypt v0.12.1
   Compiling diesel v1.4.8
   Compiling darling v0.13.4
   Compiling darling v0.14.1
    Checking comrak v0.12.1
   Compiling string_cache_codegen v0.5.2
   Compiling phf_codegen v0.8.0
error: failed to run custom build command for `lemmy_utils v0.16.5 (/****/lemmy2/crates/utils)`

Caused by:
  process didn't exit successfully: `/****/lemmy2/target/debug/build/lemmy_utils-53c738f32a60fcf3/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-changed=translations/email/en.json
  cargo:rerun-if-changed=translations/email/fi.json
  cargo:rerun-if-changed=translations/email/ko.json
  cargo:rerun-if-changed=translations/email/pt.json

  --- stderr
  Error: Fmt(Os { code: 2, kind: NotFound, message: "No such file or directory" })
warning: build failed, waiting for other jobs to finish...
error: build failed
```

**Comment by dessalines** [comment]
This is a strange one. You're sure `crates/utils/translations/email/` has those files?

Does the docker install have the same issue? If it doesn't, then it has to be your rust install. Otherwise CI would be failing.

**Comment by ahangarha** [comment]
I see the files in the folder:
```
➜  lemmy git:(main) ✗ l crates/utils/translations/email 
total 28K
drwxrwxr-x 2 sama sama 4.0K Aug  2 17:05 .
drwxrwxr-x 4 sama sama 4.0K Aug  2 17:05 ..
-rw-rw-r-- 1 sama sama 1.8K Aug  2 17:05 en.json
-rw-rw-r-- 1 sama sama 1.9K Aug  2 17:05 fi.json
-rw-rw-r-- 1 sama sama 1.2K Aug  2 17:05 it.json
-rw-rw-r-- 1 sama sama 1.8K Aug  2 17:05 ko.json
-rw-rw-r-- 1 sama sama 1.8K Aug  2 17:05 pt.json
```
I updated rust to 1.62.1 and build process could continue.

**Comment by Nutomic** [comment]
> My rust version is 1.59.0.

I just tried to build with that version and it worked without any problems. Which OS do you use?

**Comment by ahangarha** [comment]
> Which OS/distro do you use?
Kubuntu 22.04.

Then perhaps cargo needed to be upgraded. I don't know its version that time.

**Comment by jayabie** [comment]
Hello Friends, @Nutomic @ahangarha 

I just wanna check if there is any update?

**Comment by ahangarha** [comment]
Hi. I couldn't run run the project locally :(

**Comment by Nutomic** [comment]
@ahangarha Please give some more details, like the commands you ran and errors you get.

**Comment by ahangarha** [comment]
I think I had issue in connecting font and back. If you have some time, maybe we can have a short video call and fix it. Then I should be able to work.

**Comment by Nutomic** [comment]
I dont like video calls but we can chat via matrix.org, my username is `@nutomic:matrix.org`

**Comment by dessalines** [comment]
[Here's the dev chatroom](https://matrix.to/#/!tUnhsBRCsiePXfsIGe:matrix.org).

**Comment by dessalines** [comment]
Looks like someone made an RTL Arabic liemmy! https://www.dwanya.co/

Could someone who speaks Arabic contact that site so we can get their changes incorporated into lemmy-ui?

**Comment by jayabie** [comment]
@dessalines 

I do speak Arabic, but i couldn't find anything to use to reach them.

**Comment by dessalines** [comment]
Maybe you could sign up, and either make a post, or private message the mods? Looks like it has open signups.

**Comment by dessalines** [comment]
Apparently they paid someone to code this, and keep the source code private. They would have done much better to pay someone with RTL experience to update lemmy-ui itself, then they could still get code updates. As it they'll be stuck on old versions of lemmy-ui.

**Comment by jayabie** [comment]
Hello guys @dessalines @Nutomic ,

should i move with the signup, or we wait for the response to @Nutomic ?

Yes, after a further check found they didn't share their code.

**Comment by dessalines** [comment]
Lets wait for them to respond to nutomic first.

**Comment by xximj** [comment]
Hi guys 
Sorry to let you wait so long .. The code is available here :
https://github.com/xximj/dwanya
Works nice with v0.16.6.

I wish i've seen this conversation before working on the code , very interesting information shared by @ahangarha. Thanks for all your efforts guys.

**Comment by dessalines** [comment]
Thanks! Again I personally don't have time to work on this, but hopefully that code can be helpful to someone who does.

**Comment by jayabie** [comment]
No update made to this? we need the RTL from lemmy`s master branch very much so we can get the updates.

**Comment by Nutomic** [comment]
If you want this to be implemented, you can work on it yourself (or hire someone to work on it), and then make a pull request. We would definitely like to merge this functionality.

**Comment by jayabie** [comment]
No progress made on this @dessalines @Nutomic ?

**Comment by dessalines** [comment]
No. Read the comments above.

**Comment by jayabie** [comment]
I couldn't find anyone to make it this is why i'm asking.

**Comment by xximj** [comment]
Hey @jayabie 
Feel free to say hi at @mjxtb:matrix.org - we can team up on this.

**Comment by jayabie** [comment]
@xximj Thanks for your invitation, i will join soon.

**Comment by lorem10** [comment]
Hi guys, we had a draft PR that was supposed to work on other sections last week. is it in progress, I also want to contribute to this case

**Comment by hamidrezakp** [comment]
> Hi guys, we had a draft PR that was supposed to work on other sections last week. is it in progress, I also want to contribute to this case

Hey, do you mean #1963  or something else you were working on?
Anyways, we love to see progress on this part, especially on comments section.

Last week i have done some testing with comments section and it's kinda impossible to make HTML `ul`s bidirectional. only one direction can be chosen.(maybe i wasn't trying hard).

@dessalines
When can we expect the new UI to be ready?
Is it worth it to make effort and fix this UI?

**Comment by ahangarha** [comment]
For lists, only we can add direction to `ul` and not to `li`. Don't waste time on fixing issues with `li`.

**Comment by ahangarha** [comment]
May you share some screenshots of what you see as ugly?

**Comment by hamidrezakp** [comment]
> May you share some screenshots of what you see as ugly?

I couldn't make it happen, it was hypothetically.
one comment on left and another on right, inside chain of replies, should be hard to follow.

**Comment by ahangarha** [comment]
That is the nature of the issue. It happens the same if we do it here. See:

این متن فارسی است و در به صورت یک متن RTL با تراز راست‌چین نمایش داده می‌شود.

**Comment by hamidrezakp** [comment]
Any updates on this? 
@ahangarha  did you get to work on comments section?
At least we can merge the parts that is working(#1963)

**Comment by ahangarha** [comment]
I haven't worked on it yet.

Since the current partial solution doesn't harm the project, I would say we can ask for merging #1963

**Comment by hamidrezakp** [comment]
Let's have this issue open as it it not totally completed, but partially.
The comments section needs BIDI too.

**Comment by ahangarha** [comment]
@hamidrezakp Can you make a checklist of what is done and what are remaining?

**Comment by hamidrezakp** [comment]
> @hamidrezakp Can you make a checklist of what is done and what are remaining?

Sure:
- [ ] Post Title
- [x] Post Text (#1963)
- [ ] Comments

Although we should discuss if we want to have post title be bidi even if users layout is different than title direction (e.g. User with `en_us` language opens a Persian post with Persian title).

Other than post title, imo we just need comments to be bidi for now.

And please update your first comment and mention current status.

**Comment by ahangarha** [comment]
> if we want to have post title be bidi even if users layout is different than title direction
We should have bidi by I think what you mean is that if we should enforce text alignment with  `left|right` rather than `start`.

This is important but it is a design decision.

**Comment by Nutomic** [comment]
I just submitted a backend PR which allows Arabic (and Cyrillic) in usernames and community names: https://github.com/LemmyNet/lemmy/pull/4083
